### PR TITLE
Improved telemetry 4.2

### PR DESCRIPTION
**Pull request by erosb**
Same changes as in #17422 , now targeting the `master` branch.

Copying description:

# GSoC 2020 Work:

The original commits can be found in the improved-telemetry branch, and this is a squash of those.
Along with the minimal information of hazelcast version, java version, operating system, uptime, memory used, Number of clients, Number of clusters, etc. the Phonehome process now also collects the following information-

 *   Count of hazelcast maps with backup read enabled, map store, eviction, at least one query cache, hot restart enabled, at least one index, WAN replication configured, and at least one custom attribute.
 *   Count of various distributed data structures including sets, queues, multimaps, caches, list, ring buffer, topic, replicated maps, cardinality estimator, PN counter, Flake Id generator.
 *   Throughput and Latency of MapLoaders and MapStores.
 *   Details of the Cloud used for deployment of hazelcast (AWS, Azure, or GCP), also whether it runs in Kubernetes or Docker.

Refactored the initial phonehome code with some necessary changes. Created a global node in the phonehome as the four methods(check() , phoneHome() , createParameters() and addClientInfo()) of the class received same node parameter.To avoid the complications in the PhoneHome class, a MetricsCollector Interface is used, which the various collectors implemented to collect relevant metrics. A common function of computeMetrics(Node node) to collect the metric details in Phonehome, made the design optimal to use.

Moving onto the integration tests for Phonehome using wire mock, it turned out that the network communication performed by PhoneHome can not be tested. Although if we fake the call home server with wire mock, the endpoint was hardcoded in the phonehome.So to achieve testability, the URL was made parametrizable. Also, to test the latency of maps with a map store, a fake map store(Delay Map Store) is being used, which gives a delay of 200ms to achieve testability. To achieve testability for the cloud details of Hazelcast, we made the CloudInfoCollector parametrized because we had to hardcode the endpoints of the cloud service providers.
PhoneHomeBenchmark measures the average time by the phonehome process for a node using quite a significant number of data structures(10k maps with different cases and 5k other distributed objects).
PhoneHomeMetrics enum contains all the details about the keys used for the various metrics for ease.
Following is the list of new metrics that are now being collected by Phonehome-

Metrics name (Key)

  *    Number of Maps (mpct)
  *    Number of caches (cact)
  *    Number of caches with WAN replication configured (cawact)
  *    Number of maps with WAN replication configured (mpwact)
  *    Number of queues (quct)
  *    Number of multimaps (mmct)
  *    Number of sets (sect)
  *    Number of lists (lict)
  *    Number of ring buffers (rbct)
  *    Number of maps with backup read enabled (mpbrct)
  *    Number of maps using map store (mpmsct)
  *    Number of maps using at least one index (mpaoict)
  *    Number of maps having at least one custom attribute (mpaocct)
  *    Number of maps with at least one query cache (mpaoqcct)
  *    Number of maps with hotstart enabled (mphect)
  *    Number of maps with eviction (mpevct)
  *    Number of maps with native in memory (mpnmct)
  *    Number of topics (tpct)
  *    Number of replicated Map (rpct)
  *    Number of cardinality estimator (cect)
  *    Number of FlakeIdgenerator (figct)
  *    Number of PNcounter (pncct)
  *    Average Put Latency of map with mapstore (mpptlams)
  *    Average Put Latency of map without mapstore(mpptla)
  *    Average Get Latency of map with mapstore (mpgtlams)
  *    Average Get Latency of map without mapstore (mpgtla)
  *    Cloud used for deployment of Hazelcast ( cld is G it means GCP, if A means AWS and if Z it means Azure) and whether use  docker( dck is K if in Kubernetes, if Only docker then it is D and is N if it is not in a docker).

EE: https://github.com/hazelcast/hazelcast-enterprise/pull/3914

### CODE REVIEW

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/2962/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/2962/Hazelcast-pr-builder-2962.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   PhoneHomeIntegrationTest.testCacheMetrics ? FatalStartup java.lang.RuntimeExce...
[ERROR]   PhoneHomeIntegrationTest.testCountDistributedObjects ? FatalStartup java.lang....
[ERROR]   PhoneHomeIntegrationTest.testDockerStateIfKuberNetes ? FatalStartup java.lang....
[ERROR]   PhoneHomeIntegrationTest.testDockerStateIfOnlyDocker ? FatalStartup java.lang....
[ERROR]   PhoneHomeIntegrationTest.testForCloudIfAWS ? FatalStartup java.lang.RuntimeExc...
[ERROR]   PhoneHomeIntegrationTest.testForCloudIfAzure ? FatalStartup java.lang.RuntimeE...
[ERROR]   PhoneHomeIntegrationTest.testForCloudIfGCP ? FatalStartup java.lang.RuntimeExc...
[ERROR]   PhoneHomeIntegrationTest.testMapLatenciesWithMapStore ? FatalStartup java.lang...
[ERROR]   PhoneHomeIntegrationTest.testMapLatenciesWithoutMapStore ? FatalStartup java.l...
[ERROR]   PhoneHomeIntegrationTest.testMapMetrics ? FatalStartup java.lang.RuntimeExcept...
[INFO] 
[ERROR] Tests run: 35421, Failures: 0, Errors: 10, Skipped: 979
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

File: hazelcast/src/main/java/com/hazelcast/internal/util/phonehome/PhoneHome.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+import static java.lang.System.getenv;
+
+/**
+ * Pings phone home server with cluster info daily.
+ */
+public class PhoneHome {
+
+    private static final String FALSE = "false";
+    private static final String DEFAULT_BASE_PHONE_HOME_URL = "http://phonehome.hazelcast.com/ping";
```
Comment (kwart) [inline hazelcast/src/main/java/com/hazelcast/internal/util/phonehome/PhoneHome.java:38]: Can we switch to HTTPS please?

Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/internal/util/phonehome/PhoneHome.java:38]: Do we have https for the service? cc @erosb

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed.
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] COMPILATION ERROR : 
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[22,15] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:2.5.1:compile (default-compile) on project hazelcast: Compilation failure: Compilation failure: 
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[22,15] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[22,15] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsRegistryImpl.java:[29,48] error: cannot find symbol
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[22,15] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[22,15] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsRegistryImpl.java:[29,48] error: cannot find symbol
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast/src/main/java/com/hazelcast/internal/metrics/MetricsRegistry.java:[19,48] error: cannot find symbol
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsRegistryImpl.java:[309,24] error: cannot find symbol
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast/src/main/java/com/hazelcast/internal/metrics/MetricsRegistry.java:[298,17] error: cannot find symbol
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast/src/main/java/com/hazelcast/client/impl/statistics/ClientStatisticsService.java:[32,48] error: cannot find symbol
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/PublisherMetricsCollector.java:[21,48] error: cannot find symbol
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/PublisherMetricsCollector.java:[29,50] error: cannot find symbol
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/CompositeMetricsCollector.java:[20,48] error: cannot find symbol
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/CompositeMetricsCollector.java:[22,50] error: cannot find symbol
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/CompositeMetricsCollector.java:[24,18] error: cannot find symbol
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/CompositeMetricsCollector.java:[26,37] error: cannot find symbol
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast/src/main/java/com/hazelcast/client/impl/statistics/ClientStatisticsService.java:[383,23] error: cannot find symbol
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsService.java:[22,48] error: cannot find symbol
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsService.java:[159,24] error: cannot find symbol
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast/src/main/java/com/hazelcast/internal/diagnostics/MetricsPlugin.java:[21,48] error: cannot find symbol
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast/src/main/java/com/hazelcast/internal/diagnostics/MetricsPlugin.java:[85,57] error: cannot find symbol
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsCollectionCycle.java:[28,48] error: cannot find symbol
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsCollectionCycle.java:[51,18] error: cannot find symbol
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsCollectionCycle.java:[59,27] error: cannot find symbol
--------------------------
</pre>
</details>

**Comment by mmedenjak** [comment]
@erosb finished my review. Except for the two comments, looks good to me. I've squashed all previous commits to the first one so please just take a look at the second one.

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed.
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   PhoneHomeIntegrationTest.testCacheMetrics ? FatalStartup java.lang.RuntimeExce...
[ERROR]   PhoneHomeIntegrationTest.testCountDistributedObjects ? FatalStartup java.lang....
[ERROR]   PhoneHomeIntegrationTest.testDockerStateIfKuberNetes ? FatalStartup java.lang....
[ERROR]   PhoneHomeIntegrationTest.testDockerStateIfOnlyDocker ? FatalStartup java.lang....
[ERROR]   PhoneHomeIntegrationTest.testForCloudIfAWS ? FatalStartup java.lang.RuntimeExc...
[ERROR]   PhoneHomeIntegrationTest.testForCloudIfAzure ? FatalStartup java.lang.RuntimeE...
[ERROR]   PhoneHomeIntegrationTest.testForCloudIfGCP ? FatalStartup java.lang.RuntimeExc...
[ERROR]   PhoneHomeIntegrationTest.testMapLatenciesWithMapStore ? FatalStartup java.lang...
[ERROR]   PhoneHomeIntegrationTest.testMapLatenciesWithoutMapStore ? FatalStartup java.l...
[ERROR]   PhoneHomeIntegrationTest.testMapMetrics ? FatalStartup java.lang.RuntimeExcept...
[INFO] 
[ERROR] Tests run: 35491, Failures: 0, Errors: 10, Skipped: 979
[INFO] 

[ERROR] There are test failures.
</pre>
</details>

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed.
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There are 2 errors reported by Checkstyle 8.38 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/test/java/com/hazelcast/internal/util/phonehome/PhoneHomeIntegrationTest.java:35:8: Unused import - com.hazelcast.test.HazelcastParallelClassRunner. [UnusedImports]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/test/java/com/hazelcast/internal/util/phonehome/PhoneHomeIntegrationTest.java:38:8: Unused import - com.hazelcast.test.annotation.ParallelJVMTest. [UnusedImports]
--------------------------
</pre>
</details>

**Review by emre-aydin — APPROVED** [review]

File: hazelcast/src/main/java/com/hazelcast/internal/util/phonehome/PhoneHome.java (hunk 1 above)
Comment (emre-aydin) [reply hazelcast/src/main/java/com/hazelcast/internal/util/phonehome/PhoneHome.java:39]: AFAIK we currently don't have it. We can handle it separately. We can force redirect to https once that's done for older IMDG versions.

Comment (kwart) [reply hazelcast/src/main/java/com/hazelcast/internal/util/phonehome/PhoneHome.java:39]: I would like to see it resolved before merging this PR. We shouldn't leak even more info about users environment.
We can start with both endpoints available (http, https) - i.e. no redirect is needed, but new Hazelcast releases should go to the encrypted one.

Comment (erosb) [reply hazelcast/src/main/java/com/hazelcast/internal/util/phonehome/PhoneHome.java:39]: I created a callhome issue to track this: https://github.com/hazelcast/callhome/issues/35

File: hazelcast/src/main/java/com/hazelcast/internal/util/phonehome/BuildInfoCollector.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+class BuildInfoCollector implements MetricsCollector {
+
+    @Override
+    public Map<PhoneHomeMetrics, String> computeMetrics(Node hazelcastNode) {
+        BuildInfo build = BuildInfoProvider.getBuildInfo();
+        Map<PhoneHomeMetrics, String> buildInfo = new HashMap<>();
+        JetBuildInfo jetBuildInfo = hazelcastNode.getBuildInfo().getJetBuildInfo();
+        buildInfo.put(PhoneHomeMetrics.HAZELCAST_DOWNLOAD_ID, getDownloadId());
+        buildInfo.put(PhoneHomeMetrics.CLIENT_ENDPOINT_COUNT,
```
[the remark this answers is not in the document]
Comment (erosb) [reply hazelcast/src/main/java/com/hazelcast/internal/util/phonehome/BuildInfoCollector.java]: Good point, I moved it to `ClientInfoCollector` in fce44b3ed9b85ec2416ae2f857995d7a689db64f

**Review by mmedenjak — APPROVED** [review]

**Comment by Holmistr** [comment]
@erosb Just to make sure, is this part of the `master` branch (thus 5.x line) as well?

**Comment by erosb** [comment]
@Holmistr yes it is on `master`

**Comment by Holmistr** [comment]
Perfect, thanks!

**Pull request merged**